File: amazon_interview_2024_java/two_sum.java
import java.util.HashSet;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Vector;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int[] result =  new int [2];

        for (int i = 0; i < nums.length; i++) {

            int diff = target - nums[i];

            if (map.containsKey(diff)) {
                result[0] = i;
                result[1] = map.get(diff);
                return result;
            }

            if (!map.containsKey(nums[i])) {
                map.put(nums[i], i);
            }
        }

        return result;
    }
}

File: spaced_rep_june_2024/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let map = new Map(), n = nums.length

    for (let i = 0; i < n; i++) {
        if (map.has(target - nums[i])) {
            return [i, map.get(target - nums[i])]
        } else {
            map.set(nums[i], i)
        }
    }  
};

console.log(twoSum(nums = [2,7,11,15], target = 9))